### DISCUSSION TITLE
Custom Appender for Logback

### DISCUSSION THREAD

**Discussion by MihaiBeigar** [playframework/playframework#12478] [Q&A] [answered]
Using Play Framework 3.0.1
I'm trying to make a new appender for logback the thing is the error it gives makes no sense to me right now: 

```
INFO in ch.qos.logback.core.model.processor.AppenderModelHandler - About to instantiate appender of type [logback.appenders.TestAppender]
ERROR in ch.qos.logback.core.model.processor.AppenderModelHandler - Could not create an Appender of type [logback.appenders.TestAppender]. ch.qos.logback.core.util.DynamicClassLoadingException: Failed to instantiate type logback.appenders.TestAppender
Caused by: java.lang.ClassNotFoundException: logback.appenders.TestAppender
ERROR in ch.qos.logback.core.model.processor.DefaultProcessor@30720312 - Failed to traverse model appender ch.qos.logback.core.model.processor.ModelHandlerException: ch.qos.logback.core.util.DynamicClassLoadingException: Failed to instantiate type logback.appenders.TestAppender
    at ch.qos.logback.core.model.processor.ModelHandlerException: ch.qos.logback.core.util.DynamicClassLoadingException: Failed to instantiate type logback.appenders.TestAppender
```

I think that is the most important part of the Error.

What I did is created **logback.appenders.TestAppender** in the **app** folder.

The TestAppender code is pretty simple:

```java
    package logback.appenders;

    import ch.qos.logback.classic.spi.ILoggingEvent;
    import ch.qos.logback.core.AppenderBase;

    public class TestAppender extends AppenderBase<ILoggingEvent> {

        @Override
        protected void append(ILoggingEvent event) {
            if (!isStarted()) {
                return;
            }

            System.out.println( event.getFormattedMessage() );
        }

    }
```

And in the logback.xml config I added inside **configuration** tag:

```xml
    <import class="logback.appenders.TestAppender"/>

    <appender name="TEST" class="TestAppender">
    </appender>

    <root level="INFO">
      <appender-ref ref="TEST"/>
    </root>
```

If i'm not wrong everything in the app directory should be in the classpath as it gets compiled in the target/scala-2.13 directory and I've confirm that I can see the TestAppender.class inside it.

What am I missing to make this work ? 
Sorry if this is not a play specific problem, but I also search for appender examples and are very close to what I did so i thought this might have something to do with play.

**Comment by ihostage** [comment, marked as answer]
Hi, @MihaiBeigar!
Did you catch this error in Dev or Prod mode? It's can be related to specific Play classloaders that is used in Dev-mode.
To check my assumption, you can move your appender into separated library and use this library as a dependency like Logback itself.

**Reply by MihaiBeigar** [reply]
Thank you for your answer! I'm going to test this as soon as possible. By the way it was on Dev mode.

**Reply by MihaiBeigar** [reply]
Hey, @ihostage! 
Just tested it out, used Gradle to make a fast library (same code) called it org.testing.TestAppender but the problems seems to be the same as before, if you have any more ideas that would be great !
I'm on dev environment to clarify.

**Reply by ihostage** [reply]
Hmmmm  🤔  It's very strange... I was sure that your problem is related to this issues and using separated jar should help to you.
https://github.com/playframework/playframework/issues/8003
https://github.com/playframework/playframework/issues/8125

Can you create a small project (with separated jar) that reproduces your problem?

**Reply by MihaiBeigar** [reply]
Indeed seems strange to me too, was trying this on a new project so here you have it:
https://github.com/MihaiBeigar/Testing

**Reply by MihaiBeigar** [reply]
You were right ... got my hands on another PC today and the problem is solved there :(, so it's something with the environment on the development PC that is wrong or maybe IntelliJ, thank you for you help !
I assume this is only a development problem right ? In production mode this problem won't exist ?

**Reply by ihostage** [reply]
> I assume this is only a development problem right ? In production mode this problem won't exist ?

Yes, you are right! You can find a more detailed explanation in https://github.com/playframework/playframework/issues/8125 😉
